src/VirtualContent.js

```diff
@@ -0,0 +1,341 @@
+
+    this[_intersectionObserver] =
+        new IntersectionObserver(this[_intersectionObserverCallback]);
+    this[_mutationObserver] =
+        new MutationObserver(this[_mutationObserverCallback]);
+    this[_resizeObserver] = new ResizeObserver(this[_resizeObserverCallback]);
+    this[_estimatedHeights] = new WeakMap();
+    this[_updateRAFToken] = undefined;
+    this[_emptySpaceSentinelContainer] =
+        this.shadowRoot.getElementById('emptySpaceSentinelContainer');
+
+    this[_intersectionObserver].observe(this);
+    // Send a MutationRecord-like object with the current, complete list of
+    // child nodes to the MutationObserver callback; these nodes would not
+    // otherwise be seen by the observer.
+    //
+    // TODO: Find another way to do this or a better time. Technically, it's a
+    // violation of the "Requirements for custom element constructors and
+    // reactions".
+    //
+    // https://html.spec.whatwg.org/multipage/custom-elements.html#custom-element-conformance
+    this[_mutationObserverCallback]([
+      {
+        type: 'childList',
+        target: this,
+        addedNodes: Array.from(this.childNodes),
+        removedNodes: [],
+        previousSibling: null,
+        nextSibling: null,
+      }
+    ]);
+    this[_mutationObserver].observe(this, {childList: true});
+
+    this.addEventListener(
+        'activateinvisible', this[_onActivateinvisible], {capture: true});
+  }
+
+  [_intersectionObserverCallback](entries) {
+    for (const entry of entries) {
+      const target = entry.target;
+      const isIntersecting = entry.isIntersecting;
+
+      // Update if this element has moved into or out of the viewport.
+      if (target === this) {
+        this[_scheduleUpdate]();
+        break;
+      }
+
+      const targetParent = target.parentNode;
+
+      // Update if an empty space sentinel has moved into the viewport.
+      if (targetParent === this[_emptySpaceSentinelContainer] && isIntersecting) {
+        this[_scheduleUpdate]();
+        break;
+      }
+
+      // Update if a child has moved out of the viewport.
+      if (targetParent === this && !isIntersecting) {
+        this[_scheduleUpdate]();
+        break;
+      }
+    }
+  }
+
+  [_mutationObserverCallback](records) {
+    const estimatedHeights = this[_estimatedHeights];
+
+    for (const record of records) {
+      for (const node of record.removedNodes) {
+        if (node.nodeType === Node.ELEMENT_NODE) {
+          // Removed children should have be made visible again and we should
+          // unobserve them with the resize observer.
+          this[_resizeObserver].unobserve(node);
+          this[_intersectionObserver].unobserve(node);
+          node.removeAttribute('invisible');
+          estimatedHeights.delete(node);
+        }
+      }
+
+      for (const node of record.addedNodes) {
+        if (node.nodeType === Node.ELEMENT_NODE) {
+          // Added children should be invisible initially.
```

Comment:
This function schedules an update at the end and that update will end up either accounting for the estimated space taken by the element in an empty space sentinels and the placement of the rendered children or by actually making the element visible. `_update` handles all of the visibility changes based on layout but newly added elements are initially made invisible here so that you can insert lots of elements without them all being rendered on the next frame.